python/tests/compat/test_dict.py

```diff
@@ -0,0 +1,217 @@
+                          unit=sc.units.s)
+    ds.masks["amask"] = sc.Variable(dims=["y"],
+                                    values=[True, True, False, True, False])
+    # Note that attributes complicate things here, as they are duplicated in
+    # each entry during the conversion to dict. So for now, we leave attributes
+    # out.
+    ds_dict = sc.to_dict(ds)
+    assert sc.from_dict(ds_dict["a"]) == ds["a"]
+    assert sc.from_dict(ds_dict["b"]) == ds["b"]
+
+
+def test_dataset_from_dict():
+    ds_dict = {
+        "a": {
+            "coords": {
+                "x": {
+                    "dims": ["x"],
+                    "values": np.arange(10)
+                },
+                "y": {
+                    "dims": ["y"],
+                    "values": np.arange(5),
+                    "unit": sc.units.m
+                },
+            },
+            "masks": {
+                "amask": {
+                    "dims": ["y"],
+                    "values": [True, True, False, True, False]
+                }
+            },
+            "data": {
+                "dims": ["y", "x"],
+                "values": np.random.random([5, 10])
+            }
+        },
+        "b": {
+            "coords": {
+                "x": {
+                    "dims": ["x"],
+                    "values": np.arange(10)
+                },
+                "y": {
+                    "dims": ["y"],
+                    "values": np.arange(5),
+                    "unit": sc.units.m
+                },
+            },
+            "masks": {
+                "amask": {
+                    "dims": ["y"],
+                    "values": [True, True, False, True, False]
+                }
+            },
+            "data": {
+                "dims": ["y", "x"],
+                "values": np.random.random([5, 10]),
+                "variances": np.random.random([5, 10])
+            }
+        }
+    }
+    ds = sc.from_dict(ds_dict)
+    assert ds["a"] == sc.from_dict(ds_dict["a"])
+    assert ds["b"] == sc.from_dict(ds_dict["b"])
+
+
+def test_dataset_round_trip():
+    ds = sc.Dataset()
+    ds.coords["x"] = sc.Variable(dims=["x"], values=np.arange(10))
+    ds.coords["y"] = sc.Variable(dims=["y"],
+                                 values=np.arange(5),
+                                 unit=sc.units.m)
+    ds["a"] = sc.Variable(dims=["y", "x"], values=np.random.random([5, 10]))
+    ds["b"] = sc.Variable(dims=["y", "x"],
+                          values=np.random.random([5, 10]),
+                          variances=np.random.random([5, 10]),
+                          unit=sc.units.s)
+    ds.masks["amask"] = sc.Variable(dims=["y"],
+                                    values=[True, True, False, True, False])
+    # Note that round trip would not work if attrs are present, since they get
+    # become a per-item attribute during the conversion to dict.
+    assert ds == sc.from_dict(sc.to_dict(ds))
```

Comment:
I don't see a major problem with your test cases, but possibly missing several test cases to cover:
1. More than one dimension in your variables
2. Event style layout
3. aligned/realigned data.

Reply:
1. There are 2D variables in the tests for `DataArray` and `Dataset`. Is that enough or would you like to have a separate test specifically on a 2D variable?
2. Need to think about that one.
3. :+1:  I added support for unaligned data.

Reply:
So right now, the conversion to dict when event data is present just works. The `"values"` entry in the dict just contains a scipp event list. I could convert this to a Python list of numpy arrays, but i'm not sure this is what we want.
I left the variable `unit` and `dtype` as the scipp object instead of converting them to strings, because I think that is better (and this is also needed if you want to do a round trip with `to_dict` and `from_dict`).

That said, the round trip with event data will currently not work because the Variable constructor does not allow specifying the values as an event list as an input. You need to create an empty buffer and then add the event lists one by one, as in https://scipp.github.io/event-data/overview.html#Introduction

Reply:
So does that mean we need special case construction (alongside additional flags)? when storing event data in dicts so that we can construct them properly with your round-trip test?

Reply:
I also just noticed I forgot to support 0D variables....